### ISSUE TITLE
Plan for v3

### ISSUE THREAD

**Issue by chrisvfritz** [chrisvfritz/prerender-spa-plugin#123]
One of the big things holding this project back right now is PhantomJS, which we need to move away from eventually. @Tribex has experimented with [a more modular fork](https://github.com/Tribex/prerenderer-webpack-plugin), where a number of different renderers can be used. For v3 of this plugin, we'd like to team up to bring the more modular approach to feature parity (or close to it) of v2.

@drewlustro has been leading maintenance of this project lately and @Tribex architected the fork, so I'll let you two lead the discussion from here to discuss next steps. 🙂

**Comment by Tribex** [comment]
Oh, hi! Somehow I didn't see this issue until now.

Just about to go somewhere, so I'll try and get this off quick:
Basically what I'm thinking right now would be something along these lines:
1. Fork [prerenderer-webpack-plugin](https://github.com/Tribex/prerenderer-webpack-plugin) to make it API-compatible with `prerender-spa-plugin`, potentially adding an additional PhantomJS renderer.
2. See what internal changes might need to be made, squash bugs, and ideally, write tests.
3. Decide on whether [prerenderer](https://github.com/Tribex/prerenderer) should be merged into `prerender-spa-plugin` or remain separate and have `prerender-spa-plugin` be a wrapper around it in the same way `prerenderer-webpack-plugin` is.
4. Decide on which default renderer to switch to. (Potentially adding new ones.)
5. Determine on what the best upgrade path would be. Should `v3` contain significant breaking changes in API usage, clearly indicating the adoption of a new architecture and a new default renderer, or should it try to remain backward-compatible so as to not mess with people's existing webpack setups?
6. Actually do things.

It looks like a lot more effort than it probably would be. Except perhaps writing tests.

What are your thoughts? @drewlustro

**Comment by chrisvfritz** [comment]
> 1. Fork prerenderer-webpack-plugin to make it API-compatible with prerender-spa-plugin, potentially adding an additional PhantomJS renderer.

Sounds great to me. Could we keep it in a `v3` branch on this repo?

> 2. See what internal changes might need to be made, squash bugs, and ideally, write tests.

👍 

> 3. Decide on whether prerenderer should be merged into prerender-spa-plugin or remain separate and have prerender-spa-plugin be a wrapper around it in the same way prerenderer-webpack-plugin is.

Unless there would be significant advantages to merging it, I think it makes sense to keep it separate. Then we can mention it in the README for anyone who wants to use it for other build systems or even other use cases, like web scraping.

> 4. Decide on which default renderer to switch to. (Potentially adding new ones.)

Are all the renderers equally stable? If not, I'd probably focus on only one for the initial release, then we can add others later.

> 5. Determine what the best upgrade path would be. Should v3 contain significant breaking changes in API usage, clearly indicating the adoption of a new architecture and a new default renderer, or should it try to remain backward-compatible so as to not mess with people's existing webpack setups?

I think where there's no advantage to changing the API, we should keep it the same. However, I don't mind significant changes if there's some benefit and would welcome the removal of some options if they're no longer needed with the new system.

Either way, since this is only used in a single place in people's apps, the upgrade path will be very simple as long as we maintain good documentation.

> 6. Actually do things.

Sounds like a plan! 😄

**Comment by Tribex** [comment]
> Sounds great to me. Could we keep it in a v3 branch on this repo?
Yep, good idea.

> Unless there would be significant advantages to merging it, I think it makes sense to keep it separate. Then we can mention it in the README for anyone who wants to use it for other build systems or even other use cases, like web scraping.
In that case, where should `prerenderer`'s repository be located? Most of the under the hood systems in `prerender-spa-plugin` are handled in `preprenderer`. `prerenderer-webpack-plugin` is just a (very) thin wrapper.

> Are all the renderers equally stable? If not, I'd probably focus on only one for the initial release, then we can add others later.
Most of the renderers cover different use-cases. They all have significant advantages and disadvantages. It might be a good idea to switch to something more self-contained as the default renderer, along the lines of Phantom, but easier to install via. NPM. Something like [Nightmare](http://www.nightmarejs.org/).

> I think where there's no advantage to changing the API, we should keep it the same. However, I don't mind significant changes if there's some benefit and would welcome the removal of some options if they're no longer needed with the new system.
Sounds good.

**Comment by chrisvfritz** [comment]
> In that case, where should prerenderer's repository be located? Most of the under the hood systems in prerender-spa-plugin are handled in prerenderer. prerenderer-webpack-plugin is just a (very) thin wrapper.

Ah, I see. What are your thoughts? If it won't limit the kinds of things we can do with the Webpack plugin, I don't mind whether it's a separate repo or not. And if we make it a separate repo and then later decide we need tighter integration, we can always integrate prerenderer later.

For this decision, it sounds like the main trade-off is ease-of-development (everything could be in one place) vs modularity. Am I getting that right?

> Most of the renderers cover different use-cases. They all have significant advantages and disadvantages. It might be a good idea to switch to something more self-contained as the default renderer, along the lines of Phantom, but easier to install via NPM. Something like Nightmare.

Nightmare sounds like a great default.

**Comment by chrisvfritz** [comment]
If I'm ever slow in responding btw (more than a couple days), feel free to ping me here or on Twitter. Sometimes I get very busy, but other times notifications just get lost, despite my best efforts to wrangle them in Octobox. 😅

**Comment by Tribex** [comment]
> For this decision, it sounds like the main trade-off is ease-of-development (everything could be in one place) vs modularity. Am I getting that right?

Yep, that sounds about right. I think for now we can keep the repositories separate. 

> Nightmare sounds like a great default.

Awesome. I'll start trying to get that working then.

> If I'm ever slow in responding btw (more than a couple days), feel free to ping me here or on Twitter. Sometimes I get very busy, but other times notifications just get lost, despite my best efforts to wrangle them in Octobox. :sweat_smile:

No worries. I'm utterly swamped as well, so I'm moving _real_ slow on this.

**Comment by Tribex** [comment]
Had some issues with Nightmare (no support for processing of multiple tabs asynchronously,) so I'm using [puppeteer](https://github.com/GoogleChrome/puppeteer) instead, which hopefully will turn out to be a good choice all-around. Initial commit pushed to v3. Basically a hodgepodge of `prerenderer-webpack-plugin` and `prerender-spa-plugin`.

**Comment by chrisvfritz** [comment]
@Tribex Woohoo! 😄🎉

**Comment by Tribex** [comment]
I've gone ahead and created a milestone and added a couple issues to it. Still moving slow as molassess here.

**Comment by Tribex** [comment]
@chrisvfritz @drewlustro Starting to have some available time to work on this.

Made quite a bit of progress today, check out https://github.com/chrisvfritz/prerender-spa-plugin/tree/v3.

Documentation is slowly coming together, compatibility with v2 is fairly decently maintained, with deprecation warnings.

What do you think?

**Comment by stereobooster** [comment]
Hey! I want to tell you about [react-snap](https://github.com/stereobooster/react-snap), it has the same purpose as this project, it uses Puppeteer, does load performance optimisation, there a lot of documentation about integration with other libraries (like Redux, Vue) and description of different edge cases.

[Here is how to switch from prerender-spa-plugin to react-snap](https://github.com/stereobooster/prerender-spa-plugin/commit/ee73d39b862bc905b44a04c6eaa58e6730957819). 

[Here is a list of alternatives](https://github.com/stereobooster/react-snap/blob/master/doc/alternatives.md).

Do you think it makes sense to discontinue development of `prerender-spa-plugin` in favour of `react-snap`? Do you aware of cases which are not handled by `react-snap`?

**Comment by chrisvfritz** [comment]
That's great news @Tribex! I think it'd probably make sense to keep the same logo for continuity.

@stereobooster Thanks for sharing react-snap! I haven't used it, so am not sure of the scope of what it can handle, but we'd love to add it to a new `Alternatives` section in the README. PR welcome!

**Comment by Tribex** [comment]
@chrisvfritz Okey-dokey!

Going to be a little busy for the rest of the week, but next week I'll try and add some fresh examples and finish getting the documentation in order.

**Comment by Tribex** [comment]
Well, it took a bit more than a week, but I've started adding tests to prerenderer, and am hoping to make some progress here soon.

...

Soon.

**Comment by chrisvfritz** [comment]
@Tribex Woohoo! Awesome work! 🎉😄

**Comment by Tribex** [comment]
@chrisvfritz Any idea how you want to handle NPM publishing?

**Comment by chrisvfritz** [comment]
@Tribex I usually just use NPM's version and publish commands. Is that what you mean?

**Comment by Tribex** [comment]
@chrisvfritz Ah, sorry, I mean how would you like to coordinate me preparing a version and you publishing it. I think we're just about ready for alpha 0.
